man/ft2nfdump.1
.\" Copyright (c) 2022, Peter Haag
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are met:
.\"
.\"  * Redistributions of source code must retain the above copyright notice,
.\"    this list of conditions and the following disclaimer.
.\"  * Redistributions in binary form must reproduce the above copyright notice,
.\"    this list of conditions and the following disclaimer in the documentation
.\"    and/or other materials provided with the distribution.
.\"  * Neither the name of the author nor the names of its contributors may be
.\"    used to endorse or promote products derived from this software without
.\"    specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
.\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd $Mdocdate$
.Dt FT2NFDUMP 1
.Os
.Sh NAME
.Nm ft2nfdump
.Nd convert flow-tools files to nfdump format.
.Sh SYNOPSIS
.Nm 
.Op Fl r Ar ftfile
.Fl w Ar nffile
.Op Fl z=<compress>
.Op Fl j
.Op Fl y
.Op Fl E
.Op Fl c Ar num
.Op Fl h
.Op Fl V
.Sh DESCRIPTION
.Nm
is the flow-tools converter program to read, convert and store netflow data, created
with the flow-tools package. It works either as a pipe filter or reads flow-tools data
directly from a file and exports nfdump data to a file.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl r Ar ftfile
Read the flow-tools netflow data from file
.Ar ftfile
If this argument is missing,
.Nm
expects the data at
.Ar stdin
(pipe filter)
.It Fl w Ar nffile
Writes netflow data to
.Ar nffile
.It Fl z=lzo
Compress flow files with LZO1X-1 compression. Fastest compression.
.It Fl z=bz2
Compress flow files with bz2 compression. Slow but most efficient. It is not recommended 
to use bz2 in a real time capturing.
.It Fl z=lz4[:level]
Compress flow files with LZ4 compression. Fast and efficient. Optional level should be between 1..10
Changing the level results in smaller files but uses up more time to compress. 
.It Fl z=zstd[:level]
Compress flow files with ZSTD compression. Fast and efficient. Optional level should be between 1..10
Changing the level results in smaller files but uses up more time to compress. 
.It Fl j
Legacy shorthand for
.Fl z=bz2 .
.It Fl y
Legacy shorthand for
.Fl z=lz4 .
.It Fl c Ar num
Limit the number of flows to the first
.Ar num
flow records.
.It Fl E
Additionally print flows to
.Ar stdout
This option allows you to check for the correct flow conversion and should not be needed
for normal operation.
.It Fl V
Print ft2nfdump version and exit.
.It Fl h
Print help text to stdout and exit.
.El
.Sh EXIT STATUS
.Nm
returns 0 on success and 255 otherwise.
.Sh EXAMPLES
To convert a flow tools file into nfdump format:
.Dl % ft2nfdump -r <flow-tools-file> -z=lz4 -w <nfdump-file.nf>
.Dl % nfdump -r <nfdump-file.nf>
.Sh SEE ALSO
.Xr nfdump 1


ChangeLog
- 972d1fb 2026-04-28  (HEAD -> devel_1.8.x, origin/devel_1.8.x) Update Readme.md
- 8fa4fee 2026-04-28  Update Readme.md
- 50d99c6 2026-04-28  Refine and improve test suite
- df9c1a8 2026-04-28  Modify testing
- a4e49cf 2026-04-28  Cleanup includes
- 3118e88 2026-04-28  Remove compat_1.6 files
- 448166e 2026-04-27  Adapt nfreader to nffileV3 and recordV4
- 1e8a820 2026-04-27  Fix RenameAppend()
- 6b913f7 2026-04-27  Make ident fixlen and adapt tests
- 3c867c0 2026-04-26  Fix some test files
- b7d60bb 2026-04-26  Fix nfreplay bug and add nfcapd test script
- 7431904 2026-04-25  Add tor node role. Still report only exit nodes
- bd43350 2026-04-25  Fix comments
- f0d61d2 2026-04-25  Update tor code with IPv6 exit nodes
- 36a555c 2026-04-25  Cleanup torlookup code
- e50dd3f 2026-04-25  Review maxmind code and update minor issues
- 5a3ae2b 2026-04-24  Improve -G/-H handling with .flat cache file.. Add flatpath option in config file
- 4cb03fc 2026-04-24  Fix payload size in convertV2
- 396a00b 2026-04-24  Improve maxmind lookup - add flat cache
- 1ee2ff3 2026-04-24  Improve tor lookup
- 62eda08 2026-04-24  Cleanup maxmind geolookup and tor torlookup
- ec67890 2026-04-24  Cleanup lauchner and repeater
- 43a3353 2026-04-23  Adapt and improve  nflowcache.c for aggregation to v4 record
- 2fcfa54 2026-04-23  Cleanup custom aggregation for new v4 record
- 9e63106 2026-04-22  Re-enable ChangeIdent
- 1642d9c 2026-04-22  Re-enable sat_record printing
- af0d1a2 2026-04-22  Fix spelling errors
- 4827cd4 2026-04-22  Fix minor bugs
- a985f62 2026-04-11  Work on more fixes. Add entropy to payload printing in output_raw
- d5ced32 2026-04-10  Work on XXX issues
- 9f3af9a 2026-04-10  Work on filter
- 0bc6c33 2026-04-10  Adapt legacy programs to new layout
- d023b74 2026-04-08  Fixes from debug session
- 73e5bf2 2026-04-08  Add nbar array records and if/vrf records
- 276d231 2026-04-08  Free converter exporter tables
- cbf5c87 2026-04-07  Fix some bugs in exporter conversion and some cleanup
- 41e8989 2026-04-07  Add v2 exporter/sampler info to V2 converter
- e190f50 2026-04-06  Add v3 converter and fix v1/v5 collector
- cfebf09 2026-04-05  Fix bug and add stub  for 1.7.x conversion
- 4a80f2e 2026-04-05  Add missing new files
- d3aa73c 2026-04-05  Implemented nffile version 3
- fc6c5ff 2026-03-30  Enable extensions in grammar.y
- 0cd2fb1 2026-03-30  Replace all gettimeofday by clock_gettime. cleanup nfxv3.h includes
- a91643f 2026-03-30  Update all programs to v4 record
- 50595b7 2026-03-29  Update sfcapd to new v4 record
- 9fd446d 2026-03-29  - Pipeline: replace rank offset with extension offset. Fill offset table at L_END and not at start. - Netflow v5. Check at runtime to add optional extension. Prevent wasting space.
- b3ad103 2026-03-28  Status 20260328-1
- 24daa81 2026-03-26  Status 20260327-1
- d9504e5 2026-03-24  Switch to devel_1.8.x branch. Add new record v4
- 6b96df6 2026-03-04  (origin/hackathon2026) Fix proper EXnull handling and printing and update tests
- 5927ed3 2026-03-04  Fix bugs
- b29f71d 2026-03-03  Replace old number scaling code
- ba94176 2026-03-02  Avoid expensive stat() calls. Replaced by fstatat()
- ff26f32 2026-03-02  New code base for nfexpire
- 732b1e4 2026-02-28  Refrag code - move logging functions into new file logging.[ch]
- 1596835 2026-02-26  Fix other thread unsafe glibc library calls
- 5b379bf 2026-02-26  Replace all localtime() calls with localtime_r()
- e36b865 2026-02-26  Fix nfprofile and some localtime() issues
- 28b42e3 2026-02-25  Fix SetupPath()
- a59e770 2026-02-25  New bookkeeper implemented
- aac815d 2026-02-25  Remove old expire code in collector
- 13f63b0 2026-02-24  Update sflow collector to nfcapd level
- ab39ba7 2026-02-23  Updated launcher code in nfcapd
- d750f97 2026-02-23  Ready new repeater code in nfcapd
- f4c9f53 2026-02-20  Replaced run() loop in nfcapd
- d4a19a6 2026-02-17  split backend data in new nffile backend context
- 4d81212 2026-02-16  Remove old and no longer used libpcap life code
- 6c6c87c 2026-02-16  Fix #667 duration bug in -s stat output
- 14d0b46 2026-02-16  Cleanup TPACKETV3 code
- 84f3b22 2026-02-16  Cleanup BPF code. Make buffer size configurable
- fce61d7 2026-02-16  Fix IP fragment init code
- 361bd5e 2026-02-15  Fix Statistic on BPF socket
- 09de43b 2026-02-15  Fix nfpcapd build options and remove useless FIX_INCLUDES
- dbdc2ba 2026-02-14  Fix potential swap32() double definition
- b3e7370 2026-02-14  Add expire function to fragment list in nfpcapd
- 8dfa9e5 2026-02-14  Update Readme
- b9438dd 2026-02-14  Update nfpcapd - Refactor messy ProcessPacket code with state processing
- e7af66f 2026-02-12  Update nfpcapd - Improve packet processing. Make it more robust
- 9d585f8 2026-02-10  Update debug messages
- 8c29f61 2026-02-10  Replace verbose option -v to verbose level -v <n> for most binaries. Needs more work
- 3ea8d71 2026-02-09  Update nfpcapd - remove unused interrupt handler and use atomic done variables
- 9a14047 2026-02-09  Fix statistic for pcap file processing
- a2b1eb7 2026-02-09  Update nfpcapd - replace slow packet dump code by native calls
- 5ab715f 2026-02-05  Add missing packet_pcap_file.c
- 183f045 2026-02-05  Update nfpcapd - Replace pcap reader code with faster mmap version
- ac07fa1 2026-02-04  Silently skip unknown records in nfanon. Fixed #665
- 5aedf78 2026-02-04  Update nfpcapd - Modify slab allocator
- 3490191 2026-02-03  Update nfpcapd - replace rb-tree with an open hash. Improves nfpcapd performance
- dcabdb6 2026-02-01  Update nfpcapd - shrink slab check
- 0cebb7c 2026-02-01  Update nfpcapd - Fix verbose logging
- 8360a58 2026-02-01  Update nfpcapd - Update help text and man page for node cache size
- 95b252e 2026-02-01  Update nfpcapd - Replace ever growing node cache with an adaptive slab allocator
- edd6784 2026-01-31  Fix #662. Update man page
- 29cf020 2026-01-29  Update nfpcapd - Rework a lot of code. Remove lock for NewNode() in hot path

src/ft2nfdump/ft2nfdump.c
/*
 *  All rights reserved.
 *  Copyright (c) 2009-2026, Peter Haag
 *  Copyright (c) 2004-2008, SWITCH - Teleinformatikdienste fuer Lehre und Forschung
 *  Copyright (c) 2001 Mark Fullmer and The Ohio State University
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions are met:
 *
 *   * Redistributions of source code must retain the above copyright notice,
 *	   this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above copyright notice,
 *	   this list of conditions and the following disclaimer in the documentation
 *	   and/or other materials provided with the distribution.
 *   * Neither the name of the author nor the names of its contributors may be
 *	   used to endorse or promote products derived from this software without
 *	   specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
 *
 *  Flow-Tools related code taken from flow-tools-0.67 created by Mark Fullmer
 *
 */

#include <errno.h>
#include <fcntl.h>
#include <netinet/in.h>
#include <stdarg.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/param.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/uio.h>
#include <time.h>
#include <unistd.h>

#include "barrier.h"
#include "ftlib.h"
#include "nfdump.h"
#include "nffile.h"
#include "nfxV3.h"
#include "output_short.h"
#include "util.h"
#include "version.h"

/* Global defines */
#define MAXRECORDS 30

static bool HasFlows = false;

typedef struct v5_block_s {
    uint32_t srcaddr;
    uint32_t dstaddr;
    uint32_t dPkts;
    uint32_t dOctets;
    uint8_t data[4];  // link to next record
} v5_block_t;

/* externals */
extern uint32_t Max_num_extensions;

/* prototypes */
void usage(char *name);

static int flows2nfdump(struct ftio *ftio, char *wfile, int compress, int extended, uint32_t limitflows);

#include "nffile_inline.c"

void usage(char *name) {
    printf(
        "usage %s [options] \n"
        "-h\t\tthis text you see right here.\n"
        "-E\t\tDump records in ASCII extended format to stdout.\n"
        "-c <num>\tLimit number of records to convert.\n"
        "-V\t\tPrint version and exit.\n"
        "-r <file>\tread flow-tools records from file\n"
        "-w <file>\twrite nfdump records to file\n"
        "-z=lzo\t\tLZO compress flows in output file.\n"
        "-z=bz2\t\tBZIP2 compress flows in output file.\n"
        "-z=lz4[:level]\tLZ4 compress flows in output file.\n"
        "-z=zstd[:level]\tZSTD compress flows in output file.\n"
        "-j\t\tLegacy shorthand for -z=bz2.\n"
        "-y\t\tLegacy shorthand for -z=lz4.\n"
        "Convert flow-tools format to nfdump format:\n"
        "ft2nfdump -r <flow-tools-data-file> -w <nfdump-file> [-z]\n",
        name);

}  // End of usage

static uint16_t *GenExtensionList(struct ftio *ftio, uint32_t *extensionSize, uint32_t *numExtensions) {
    int i;

// maximux of extensions + terminating ExNULL
#define FTMAXEXTENSIONS 10
    uint16_t *extensionList = calloc(FTMAXEXTENSIONS, sizeof(uint16_t));
    if (!extensionList) {
        LogError("malloc() error in %s:%d: %s", __FILE__, __LINE__, strerror(errno));
        return NULL;
    }

    *extensionSize = 0;
    *numExtensions = 0;
    i = 0;
    if (!ftio_check_xfield(ftio, FT_XFIELD_SRCPORT | FT_XFIELD_DSTPORT | FT_XFIELD_TCP_FLAGS | FT_XFIELD_PROT | FT_XFIELD_UNIX_SECS |
                                     FT_XFIELD_UNIX_NSECS | FT_XFIELD_SYSUPTIME | FT_XFIELD_DOCTETS | FT_XFIELD_DPKTS)) {
        extensionList[i++] = EXgenericFlowID;
        *extensionSize += EXgenericFlowSize;
        (*numExtensions)++;
    }
    if (!ftio_check_xfield(ftio, FT_XFIELD_SRCADDR | FT_XFIELD_DSTADDR)) {
        extensionList[i++] = EXipv4FlowID;
        *extensionSize += EXipv4FlowSize;
        (*numExtensions)++;
    }
    if (!ftio_check_xfield(ftio, FT_XFIELD_INPUT | FT_XFIELD_OUTPUT | FT_XFIELD_SRC_MASK | FT_XFIELD_DST_MASK)) {
        extensionList[i++] = EXflowMiscID;
        *extensionSize += EXflowMiscSize;
        (*numExtensions)++;
    }
    if (!ftio_check_xfield(ftio, FT_XFIELD_DFLOWS)) {
        HasFlows = true;
    }
    if (!ftio_check_xfield(ftio, FT_XFIELD_SRC_AS | FT_XFIELD_DST_AS)) {
        extensionList[i++] = EXasRoutingID;
        *extensionSize += EXasRoutingSize;
        (*numExtensions)++;
    }
    if (ftio_check_xfield(ftio, FT_XFIELD_PEER_NEXTHOP) == 0) {
        extensionList[i++] = EXipNextHopV4ID;
        *extensionSize += EXipNextHopV4Size;
        (*numExtensions)++;
    }
    if (!ftio_check_xfield(ftio, FT_XFIELD_EXADDR)) {
        extensionList[i++] = EXipReceivedV4ID;
        *extensionSize += EXipReceivedV4Size;
        (*numExtensions)++;
    }
    extensionList[i] = EXnull;

    return extensionList;

}  // End of GenExtensionList

static int flows2nfdump(struct ftio *ftio, char *wfile, int compress, int extended, uint32_t limitflows) {
    // required flow tools variables
    struct fttime ftt;
    struct fts3rec_offsets fo;
    struct ftver ftv;
    char *rec;
    // nfdump variables

    char *ident = "flow-tools";
    nffile_t *nffile = OpenNewFile(wfile, CREATOR_FT2NFDUMP, compress, NOT_ENCRYPTED);
    if (!nffile) {
        LogError("OpenNewFile() failed.");
        return 1;
    }
    dataBlock_t *dataBlock = WriteBlock(nffile, NULL);

    ftio_get_ver(ftio, &ftv);
    memset((void *)&fo, 0xFF, sizeof(fo));
    fts3rec_compute_offsets(&fo, &ftv);

    uint32_t recordSize = 0;
    uint32_t numElements = 0;
    uint16_t *extensionInfo = GenExtensionList(ftio, &recordSize, &numElements);
    dbg_printf("GenExtensionList: numElements: %u, recordSize: %u\n", numElements, recordSize);
    if (numElements == 0) {
        LogError("No usable fields found it flowtools file");
        return 1;
    }
    recordSize += sizeof(recordHeaderV3_t);

    int cnt = 0;
    while ((rec = ftio_read(ftio))) {
        int i, exID;
        dbg_printf("FT record %u\n", cnt);
        if (!IsAvailable(dataBlock, recordSize)) {
            // flush block - get an empty one
            dataBlock = WriteBlock(nffile, dataBlock);
        }

        void *buffPtr = GetCurrentCursor(dataBlock);
        AddV3Header(buffPtr, recordHeader);

        // header data
        recordHeader->engineType = *((uint8_t *)(rec + fo.engine_type));
        recordHeader->engineID = *((uint8_t *)(rec + fo.engine_id));

        i = 0;
        while ((exID = extensionInfo[i]) != EXnull) {
            dbg_printf("Process slot %i extension %u - %s\n", i, exID, extensionTable[exID].name);
            switch (exID) {
                case EXgenericFlowID: {
                    uint32_t when, unix_secs, unix_nsecs, sysUpTime;
                    unix_secs = *((uint32_t *)(rec + fo.unix_secs));
                    unix_nsecs = *((uint32_t *)(rec + fo.unix_nsecs));
                    sysUpTime = *((uint32_t *)(rec + fo.sysUpTime));

                    PushExtension(recordHeader, EXgenericFlow, genericFlow);
                    when = *((uint32_t *)(rec + fo.First));
                    ftt = ftltime(sysUpTime, unix_secs, unix_nsecs, when);
                    genericFlow->msecFirst = (1000LL * (uint64_t)ftt.secs) + (uint64_t)ftt.msecs;

                    when = *((uint32_t *)(rec + fo.Last));
                    ftt = ftltime(sysUpTime, unix_secs, unix_nsecs, when);
                    genericFlow->msecLast = (1000LL * (uint64_t)ftt.secs) + (uint64_t)ftt.msecs;

                    genericFlow->inPackets = *((uint32_t *)(rec + fo.dPkts));
                    genericFlow->inBytes = *((uint32_t *)(rec + fo.dOctets));
                    genericFlow->srcPort = *((uint16_t *)(rec + fo.srcport));
                    genericFlow->dstPort = *((uint16_t *)(rec + fo.dstport));
                    genericFlow->proto = *((uint8_t *)(rec + fo.prot));
                    genericFlow->tcpFlags = *((uint8_t *)(rec + fo.tcp_flags));
                    genericFlow->srcTos = *((uint8_t *)(rec + fo.tos));
                    genericFlow->fwdStatus = 0;
                } break;
                case EXipv4FlowID:
                    PushExtension(recordHeader, EXipv4Flow, ipv4Flow);
                    ipv4Flow->srcAddr = *((uint32_t *)(rec + fo.srcaddr));
                    ipv4Flow->dstAddr = *((uint32_t *)(rec + fo.dstaddr));
                    break;
                case EXflowMiscID:
                    PushExtension(recordHeader, EXflowMisc, flowMisc);
                    flowMisc->input = *((uint16_t *)(rec + fo.input));
                    flowMisc->output = *((uint16_t *)(rec + fo.output));
                    flowMisc->srcMask = *((uint8_t *)(rec + fo.src_mask));
                    flowMisc->dstMask = *((uint8_t *)(rec + fo.dst_mask));
                    flowMisc->dir = 0;
                    flowMisc->dstTos = 0;
                    break;
                case EXasRoutingID:
                    PushExtension(recordHeader, EXasRouting, asRouting);
                    asRouting->srcAS = *((uint16_t *)(rec + fo.src_as));
                    asRouting->dstAS = *((uint16_t *)(rec + fo.dst_as));
                    break;
                case EXipNextHopV4ID:
                    PushExtension(recordHeader, EXipNextHopV4, ipNextHopV4);
                    ipNextHopV4->ip = *((uint32_t *)(rec + fo.peer_nexthop));
                    break;
                case EXipReceivedV4ID:
                    PushExtension(recordHeader, EXipReceivedV4, received);
                    received->ip = *((uint32_t *)(rec + fo.exaddr));
                    break;
            }
            i++;
        }

        // update file record size ( -> output buffer size )
        dataBlock->NumRecords++;
        dataBlock->size += recordSize;

        dbg_assert(recordHeader->size == recordSize);

        if (extended) {
            flow_record_short(stdout, recordHeader);
        }

        cnt++;
        if (cnt == limitflows) break;

    } /* while */

    SetIdent(nffile, ident);
    FlushBlock(nffile, dataBlock);
    FlushFile(nffile);
    DisposeFile(nffile);
    return 0;

}  // End of flows2nfdump

int main(int argc, char **argv) {
    struct ftio ftio;
    struct stat statbuf;
    uint32_t limitflows;
    int extended, ret, fd, compress;

    /* init fterr */
    fterr_setid(argv[0]);

    extended = 0;
    limitflows = 0;
    char *ftfile = NULL;
    char *wfile = NULL;
    compress = NOT_COMPRESSED;

    int c;
    while ((c = getopt(argc, argv, "z::jyEVc:hr:w:")) != EOF) {
        switch (c) {
            case 'h': /* help */
                usage(argv[0]);
                exit(0);
                break;
            case 'V':
                printf("%s: %s\n", argv[0], versionString());
                exit(0);
                break;
            case 'E':
                extended = 1;
                break;
            case 'c':
                limitflows = atoi(optarg);
                if (!limitflows) {
                    LogError("Option -c needs a number > 0");
                    exit(255);
                }
                break;
            case 'j':
                if (compress) {
                    LogError("Use one compression: -z for LZO, -j for BZ2 or -y for LZ4 compression");
                    exit(EXIT_FAILURE);
                }
                compress = BZ2_COMPRESSED;
                break;
            case 'y':
                if (compress) {
                    LogError("Use one compression: -z for LZO, -j for BZ2 or -y for LZ4 compression");
                    exit(EXIT_FAILURE);
                }
                compress = LZ4_COMPRESSED;
                break;
            case 'z':
                if (compress) {
                    LogError("Use one compression: -z for LZO, -j for BZ2 or -y for LZ4 compression");
                    exit(EXIT_FAILURE);
                }
                if (optarg == NULL) {
                    compress = LZO_COMPRESSED;
                    LogInfo("Legacy option -z defaults to -z=lzo. Use -z=lzo, -z=lz4, -z=bz2 or z=zstd for valid compression formats");
                } else {
                    int ret = ParseCompression(optarg);
                    if (ret == -1) {
                        LogError("Usage for option -z: set -z=lzo, -z=lz4, -z=bz2 or z=zstd for valid compression formats");
                        exit(EXIT_FAILURE);
                    }
                    compress = (unsigned)ret;
                }
                break;
            case 'r':
                ftfile = optarg;
                if ((stat(ftfile, &statbuf) < 0) || !(statbuf.st_mode & S_IFREG)) {
                    LogError("No such file: '%s'", ftfile);
                    exit(255);
                }
                break;
            case 'w':
                CheckArgLen(optarg, MAXPATHLEN);
                wfile = optarg;
                break;

            default:
                usage(argv[0]);
                exit(1);
                break;

        } /* switch */
    }  // End while

    if (argc - optind) fterr_errx(1, "Extra arguments starting with %s.", argv[optind]);

    if (ftfile) {
        fd = open(ftfile, O_RDONLY, 0);
        if (fd < 0) {
            LogError("Can't open file '%s': %s.", ftfile, strerror(errno));
            exit(255);
        }
    } else {
        fd = 0;
    }

    if (!wfile) {
        usage(argv[0]);
        exit(EXIT_FAILURE);
    }

    int numWorkers = GetNumWorkers(0);
    if (!Init_nffile(numWorkers, NULL)) exit(EXIT_FAILURE);

    /* read from fd */
    if (ftio_init(&ftio, fd, FT_IO_FLAG_READ) < 0) fterr_errx(1, "ftio_init(): failed");

    ret = flows2nfdump(&ftio, wfile, compress, extended, limitflows);

    return ret;

}  // End of main
